File: day14.py
import re
with open("data/input14.txt") as f:
    pattern="p=(-?\d+),(-?\d+) v=(-?\d+),(-?\d+)"
    robots=[[int(x) for x in line] for line in re.findall(pattern,f.read())]
W,H=101,103

def update(robot,steps):
    x,y,vx,vy=robot
    robot[0]=(x+steps*vx)%W
    robot[1]=(y+steps*vy)%H

for i in range(100):
    LH,LV=[0]*W,[0]*H
    Hok,Vok=False,False
    for r in robots:
        if LH[r[0]]>20:
            Hok=True
        else:
            LH[r[0]]+=1
        if LV[r[1]]>20:
            Vok=True
        else:
            LV[r[1]]+=1
        update(r,1)
    if Hok:Hresidue=i
    if Vok:Vresidue=i
    
a=len([r for r in robots if r[0]<(W-1)//2 and r[1]<(H-1)//2])
b=len([r for r in robots if r[0]<(W-1)//2 and r[1]>(H-1)//2])
c=len([r for r in robots if r[0]>(W-1)//2 and r[1]<(H-1)//2])
d=len([r for r in robots if r[0]>(W-1)//2 and r[1]>(H-1)//2])
part1=a*b*c*d
for k in range(W*H):
    if k%W==Hresidue and k%H==Vresidue:
        part2=k;break
print(part1,part2)

File: day14.lua
local part1,part2=0,0
local W,H=101,103
local Robots={}
io.input("data/input14.txt")
local inp=io.read("*all")
for x,y,vx,vy in string.gmatch(inp,"p=(%d+),(%d+) v=(%-?%d+),(%-?%d+)") do
    table.insert(Robots,{tonumber(x),tonumber(y),tonumber(vx),tonumber(vy)})
end

function update(r)
    local x,y,vx,vy=r[1],r[2],r[3],r[4]
    r[1]=(x+vx)%W
    r[2]=(y+vy)%H
end
local Hresidue,Vresidue=0,0
for i=0,99 do
    local LH,LV={},{}
    local Hok,Vok=false,false
    for _,r in ipairs(Robots) do
        if LH[r[1]] and LH[r[1]]>20 then
            Hok=true
        else
            LH[r[1]]= LH[r[1]] and LH[r[1]]+1 or 1
        end
        if LV[r[2]] and LV[r[2]]>20 then
            Vok=true
        else
            LV[r[2]]= LV[r[2]] and LV[r[2]]+1 or 1
        end
        update(r)
        if Hok then Hresidue=i end
        if Vok then Vresidue=i end
    end
end

local a,b,c,d=0,0,0,0
for _,r in ipairs(Robots) do
    if r[1]<(W-1)/2 and r[2]<(H-1)/2 then a=a+1 end
    if r[1]<(W-1)/2 and r[2]>(H-1)/2 then b=b+1 end
    if r[1]>(W-1)/2 and r[2]<(H-1)/2 then c=c+1 end
    if r[1]>(W-1)/2 and r[2]>(H-1)/2 then d=d+1 end
end
part1=a*b*c*d
for k=1,W*H do
    if k%W==Hresidue and k%H==Vresidue then
        part2=k;break
    end
end
print(part1,part2)